Composer/packages/lib/indexers/src/dialogUtils/dialogDiff.ts

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+/**
+ * not like JsonDiff, DialogDiff compaire stop at { $kind, $id, ... }
+ * The atomic unit to add/delete/update is a dialog object {}
+ */
+
+import has from 'lodash/has';
+import get from 'lodash/get';
+import isEqual from 'lodash/isEqual';
+import { SDKKinds } from '@bfc/shared';
+
+import {
+  JsonDiffAdds,
+  JsonDiffUpdates,
+  IJsonChanges,
+  IJSONChangeAdd,
+  IJSONChangeDelete,
+  IJSONChangeUpdate,
+  IComparator,
+  IStopper,
+  defualtJSONStopComparison,
+  getWithJsonPath,
+} from './jsonDiff';
```

Comment:
I think you used a lot of interfaces, showing you trying to think this as a more general level, which is a good design thinking. 

But there are two things that are as important as a generic mindset
1. if you are thinking from a generic perspective, make sure the concepts in this layer are consistent, which means, make sure they are all high-level concept or low-level concepts.  Check your abstractions, are they mixing high-level and low level concept, are they mixing interface and implementation, are we have very similar things. 
2. Both less engineering and over engineering is not good, but usually over engineering hurts more than less engineering. This is usually a result of 1.